### PR TITLE
Hantec API Ecommerce

### PR DESCRIPTION
**Pull request by JAntonioSalas**
Revisión de código

### CODE REVIEW

File: hantec_api_ecommerce/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from . import controllers
```
Comment (hugho-ad) [inline hantec_api_ecommerce/__init__.py:2]: EOF

leave and empty line at the end of file

apply it to all files please

File: hantec_api_ecommerce/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+
+{
+    "name": "Hantec API Ecommerce",
+    "summary": "Hantece API Ecommerce",
+    "version": "1.0.0",
```
Comment (hugho-ad) [inline hantec_api_ecommerce/__manifest__.py]:
```suggestion
    "version": "15.0.0.0",
```

File: hantec_api_ecommerce/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (hugho-ad) [inline hantec_api_ecommerce/__manifest__.py]:
```suggestion
```

File: hantec_api_ecommerce/controllers/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/__init__.py:1]:
```suggestion
```

File: hantec_api_ecommerce/controllers/__init__.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
...
+
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/__init__.py]:
```suggestion
```

File: hantec_api_ecommerce/controllers/main.py [hunk 6]
```diff
@@ -0,0 +1,570 @@
+
+from odoo.http import request, Response, Controller, route
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]:
```suggestion
```
move it after import logging

File: hantec_api_ecommerce/controllers/main.py [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+import json
+import logging
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]:
```suggestion
import logging
from odoo.http import Controller, request,  route
```

File: hantec_api_ecommerce/controllers/main.py (hunk 6 above)
Comment (hugho-ad) [reply hantec_api_ecommerce/controllers/main.py]: Response us not used

File: hantec_api_ecommerce/controllers/main.py [hunk 8]
```diff
@@ -0,0 +1,570 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class MainController(Controller):
+    @route('/create_contact', methods=['POST'], type='json', auth='user' )
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]:
```suggestion
    @route("/create_contact", methods=["POST"], type="json", auth="user")
```
Use double quoting please, check all coincidences

File: hantec_api_ecommerce/controllers/main.py [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+    def create_contact(self):
+        
+        # Lista de todos los campos posibles
+
+        all_fields = [
+            'company_type', 'name', 'phone', 'street_number', 'street_number2',
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: keep as a list 
```suggestion
            'company_type', 
            'name', 
            'phone', 
            'street_number', 
            'street_number2',
```

an so on, replace single quote by double quoting

File: hantec_api_ecommerce/controllers/main.py [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+            'email', 'street', 'street2', 'city', 'zip', 'state_id', 'country_id',
+            'vat', 'l10n_mx_edi_curp', 'website', 'lang', 'category_id', 'l10n_mx_edi_fiscal_regime', 'parent_id'
+        ]
+
+        # Construir el diccionario contact_data con solo los campos enviados en el JSON
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: We prefer English 
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#idioms

File: hantec_api_ecommerce/controllers/main.py [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+        contact_data = {field: request.jsonrequest.get(field) for field in all_fields if request.jsonrequest.get(field) is not None}
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]:
```suggestion
        contact_data = {
            field: request.jsonrequest.get(field) for field in all_fields if request.jsonrequest.get(field) is not None
        }
```

take care of indentation please

you can use black 
https://pypi.org/project/black/

File: hantec_api_ecommerce/controllers/main.py [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+        for key, value in contact_data.items():
+            if isinstance(value, str) and key not in list_not_title:
+                    contact_data[key] = value.title()  # Convierte la primera letra de cada palabra a mayúscula
+
+        if existing_contact:
+            # Actualizar el contacto existente si se encontró
+            existing_contact.write(contact_data)
+            logger.info("Contacto actualizado con ID %s", existing_contact.id)
+            return {"message": f"Contacto actualizado con ID: {existing_contact.id}.", "contact_id": existing_contact.id}
+        else:
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]:
```suggestion
```
 Unnecessary "else" after "return", remove the "else" and de-indent the code inside it

File: hantec_api_ecommerce/controllers/main.py [hunk 13]
```diff
@@ -0,0 +1,570 @@
...
+        for key, value in contact_data.items():
+            if isinstance(value, str) and key not in list_not_title:
+                    contact_data[key] = value.title()  # Convierte la primera letra de cada palabra a mayúscula
+
+        if existing_contact:
+            # Retornar el ID
+
+            logger.info("Contacto encontrado con ID %s", existing_contact.id)
+            return {"message": f"Contacto encontrado con ID: {existing_contact.id}.", "contact_id": existing_contact.id}
+        else:
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py:155]: Unnecessary "else" after "return", remove the "else" and de-indent the code inside it

File: hantec_api_ecommerce/controllers/main.py [hunk 14]
```diff
@@ -0,0 +1,570 @@
...
+        if invoice.l10n_mx_edi_cfdi_uuid:
+            return {"message": "La factura ya está timbrada.", "UUID": invoice.l10n_mx_edi_cfdi_uuid}
+        
+        # Intentar el proceso de timbrado
+        invoice.action_process_edi_web_services()
+        
+        # Verificar nuevamente si la factura fue timbrada exitosamente
+        if invoice.l10n_mx_edi_cfdi_uuid:
+            return {"success": "La factura ha sido timbrada con éxito.", "UUID": invoice.l10n_mx_edi_cfdi_uuid}
+        else:
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: Unnecessary "else" after "return", remove the "else" and de-indent the code inside it

File: hantec_api_ecommerce/controllers/main.py [hunk 15]
```diff
@@ -0,0 +1,570 @@
...
+                'lines': lines_data,
+                'address_data': address_data
+            }
+            shipping_data.append(shipping_info)   
+
+        shipping_data.append({
+            "sales_team": order.team_id.id,
+            "market_place_reference": order.channel_order_reference,
+            "sale_order_name": order.name,
+            "picking_id": picking.id,
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: Using possibly undefined loop variable 'picking'

picking is defined as a for loop var only

File: hantec_api_ecommerce/controllers/main.py [hunk 16]
```diff
@@ -0,0 +1,570 @@
...
+            "picking_name": picking.name
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: same

File: hantec_api_ecommerce/__manifest__.py [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+{
+    "name": "Hantec API Ecommerce",
+    "summary": "Hantece API Ecommerce",
+    "version": "1.0.0",
+    "author": "C&O PROJECTS AND SOLUTIONS",
+    "category": "Sales Managment",
```
Comment (hugho-ad) [inline hantec_api_ecommerce/__manifest__.py:7]: Missing required key "license" in manifest file

**Comment by hugho-ad** [comment]
Missing ./README.rst file. Template here: https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: hantec_api_ecommerce/controllers/main.py [hunk 18]
```diff
@@ -0,0 +1,570 @@
...
+    @route('/invoice_sale_order/<model("sale.order"):order>', methods=['POST'], type='json', auth='user')
+    def invoice_sale_order(self, order=False):
+        data = request.jsonrequest
+        context = {
+            'active_model': 'sale.order',
+            'active_ids': [order.id],
+            'active_id': order.id,
+        }
+        
+        invoice_wizard = request.env['sale.advance.payment.inv'].with_context(context).create({
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]:
```suggestion
        invoice_wizard = request.env['sale.advance.payment.inv'].with_context(**context).create({
```

File: hantec_api_ecommerce/controllers/main.py [hunk 19]
```diff
@@ -0,0 +1,570 @@
...
+            'company_type', 'name', 'phone', 'street_number', 'street_number2',
+            'email', 'street', 'street2', 'city', 'zip', 'state_id', 'country_id',
+            'vat', 'l10n_mx_edi_curp', 'website', 'lang', 'category_id', 'l10n_mx_edi_fiscal_regime', 'parent_id'
+        ]
+
+        # Construir el diccionario contact_data con solo los campos enviados en el JSON
+        contact_data = {field: request.jsonrequest.get(field) for field in all_fields if request.jsonrequest.get(field) is not None}
+
+        # Si category_id está presente, convertirlo en una lista
+        if 'category_id' in contact_data and not isinstance(contact_data['category_id'], list):
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: because category_id is a m2m field?

which error do you get if you pass the id right as it is?

File: hantec_api_ecommerce/controllers/main.py [hunk 20]
```diff
@@ -0,0 +1,570 @@
...
+            contact_data['category_id'] = [contact_data['category_id']]
+        
+        list_not_title = ['lang', 'company_type', 'email', 'website']
+
+        for key, value in contact_data.items():
+            if isinstance(value, str) and key not in list_not_title:
+                    contact_data[key] = value.title()
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: why did you change the values?

you should register as it comes, not data transforming

File: hantec_api_ecommerce/controllers/main.py [hunk 21]
```diff
@@ -0,0 +1,570 @@
...
+        data = request.jsonrequest
+        partner_id = data.get('partner_id')  # ID del contacto a actualizar
+        update_vals = data.get('update_vals')  # Valores a actualizar
+
+        # Encontrar el contacto por su ID
+        partner = request.env['res.partner'].browse(partner_id)
+
+        list_not_title = ['lang', 'company_type', 'email', 'website', 'vat']
+
+        for key, value in update_vals.items():
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: repeted code, should be this a function

File: hantec_api_ecommerce/controllers/main.py [hunk 22]
```diff
@@ -0,0 +1,570 @@
...
+    @route('/manage_contact', methods=['POST'], type='json', auth='user')
+    def manage_contact(self):
+        
+        data = request.jsonrequest
+        email = data.get('email')
+        phone = data.get('phone')
+        contact_data = data.get('contact_data', {})  # Datos del contacto a actualizar/crear
+
+        # Verificar si ya existe un contacto a partir de los primeros digitos del correo y ultimos digitos del telefono
+        email_prefix = email[0:5]
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: better apply a regex to get the email before the domain and the @

File: hantec_api_ecommerce/controllers/main.py [hunk 23]
```diff
@@ -0,0 +1,570 @@
...
+        phone_suffix = phone[len(phone)-5:len(phone)]
+
+        # Crear dominios de búsqueda usando patrones con comodines
+        domain = [('email', 'ilike', f'{email_prefix}%'), ('phone', 'ilike', f'%{phone_suffix}')]
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: why no you use the ids?

I mean, you site are creating contacts and returning it's ids to allow post identification

File: hantec_api_ecommerce/controllers/main.py [hunk 24]
```diff
@@ -0,0 +1,570 @@
...
+
+        # Actualizar el contacto con los nuevos valores
+        partner.write(update_vals)
+
+        logger.info("Contacto actualizado con ID %s", partner_id)
+
+        return {"message": f"Contacto con ID: {partner_id} actualizado exitosamente."}
+    
+    @route('/manage_contact', methods=['POST'], type='json', auth='user')
+    def manage_contact(self):
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: each endpoint has to have a method docstring please

File: hantec_api_ecommerce/controllers/main.py [hunk 25]
```diff
@@ -0,0 +1,570 @@
...
+        contact_data = data.get('contact_data', {})  # Datos del contacto a crear
+
+        # Verificar si ya existe un contacto a partir de los últimos digitos del nombre
+        store_name_suffix = store_name[len(store_name)-4:len(store_name)]
+
+        # Crear dominios de búsqueda usando patrones con comodines
+        domain = [('name', 'ilike', f'%{store_name_suffix}')]
+
+        # Realizar la búsqueda con el dominio definido
+        existing_contact = request.env['res.partner'].search(domain, limit=1)
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py]: seams duplicated endpoint, just by the store_name

File: hantec_api_ecommerce/controllers/main.py [hunk 26]
```diff
@@ -0,0 +1,570 @@
...
+      
+    @route('/address_invoice', methods=['POST'], type='json', auth='user')
+    def create_address_invoice(self):
+        env = request.env
+        data = request.jsonrequest
+        partner_id = data.get('partner_id')
+        invoice_data = data.get('invoice_data')
+
+        partner = env['res.partner'].browse(partner_id)
+
```
Comment (hugho-ad) [inline hantec_api_ecommerce/controllers/main.py:182]: avoid to many blank lines please,

**Review by hugho-ad — CHANGES_REQUESTED** [review]

File: hantec_api_ecommerce/controllers/main.py (hunk 23 above)
Comment (JAntonioSalas) [reply hantec_api_ecommerce/controllers/main.py]: Because there are already contacts in odoo that we want to know if they exist before creating them
